### PR TITLE
Feat/is 585 govt sgid login rollout

### PR DESCRIPTION
**Pull request by alexanderleegs**
## Problem

This PR adds support for govt officials to log in via singpass. To be reviewed in conjunction with PR [#1594](https://github.com/isomerpages/isomercms-frontend/pull/1594) on the isomercms-frontend repo. It changes the existing sgin-login endpoint to return an email instead of just the status, and adds an additional endpoint to allow for users with multiple emails to call a separate endpoint to login after the initial sgid login step. It also modifies the existing behaviour of the add collaborators endpoint so that new users can be added as a collaborator directly instead of requiring them to log into the cms first.

**Breaking Changes**

- [x]  Yes - this PR contains breaking changes
  - Minor breaking change - old sgid login will no longer work since the requested permissions no longer match. Acceptable behaviour because the existing sgid login is only usable by ogp officers currently

## Tests

- [ ] Run tests on associated frontend PR
- [ ] Check that a collaborator that does not previously exist can be added to a site

### CODE REVIEW

**Comment by seaerchin** [comment]
Why would a user have separate emails?

File: src/errors/SgidErrors.ts
Comment (seaerchin) [inline src/errors/SgidErrors.ts:1]: File name should be singular

File: src/routes/v2/sgidAuth.ts
Comment (seaerchin) [inline src/routes/v2/sgidAuth.ts:100]: Super nit: is it actually officer emails or officer details? If it's details then we should update the prior method call, otherwise this var name

File: src/routes/v2/sgidAuth.ts
Comment (seaerchin) [inline src/routes/v2/sgidAuth.ts:112]: No error messages for this whole chunk - could we put in something meaningful to the user? 

Otherwise, they'd be redirected with no toast on why

File: src/routes/v2/sgidAuth.ts
Comment (seaerchin) [inline src/routes/v2/sgidAuth.ts:147]: Why do we not require the same expiring and signing for single email users?

File: src/routes/v2/sgidAuth.ts
Comment (seaerchin) [inline src/routes/v2/sgidAuth.ts:175]: Should we pass the message from sgid here?

Comment (seaerchin) [reply src/routes/v2/sgidAuth.ts:175]: Or at least provide some message for Fe to consume?

File: src/routes/v2/sgidAuth.ts
Comment (seaerchin) [inline src/routes/v2/sgidAuth.ts:180]: If this is going to be repeated, we can extract as a function.

File: src/routes/v2/sgidAuth.ts
Comment (seaerchin) [inline src/routes/v2/sgidAuth.ts:197]: nit: seems like these 2 should be packaged together

File: src/routes/v2/sgidAuth.ts
Comment (seaerchin) [inline src/routes/v2/sgidAuth.ts:214]: supernit: can just return the data here or omit entirely so intention is clear

File: src/routes/v2/sgidAuth.ts
Comment (seaerchin) [inline src/routes/v2/sgidAuth.ts:254]: our path structure isn't ideal but this is something that can be fixed down the line. this should instead be `/verify/multi/login` or something

File: src/services/utilServices/SgidAuthService.ts
Comment (seaerchin) [inline src/services/utilServices/SgidAuthService.ts:104]: supernit: can just combine (not imporant to fix)

File: src/types/sgid.ts
Comment (seaerchin) [inline src/types/sgid.ts:11]: dangerous assertion - we first need an assertion of `!!item` as this doesn't guarantee existence of `item`

**Review by seaerchin — COMMENTED** [review]
minor comments

File: src/routes/v2/sgidAuth.ts
Comment (alexanderleegs) [reply src/routes/v2/sgidAuth.ts:112]: The reason why no error messages are sent is because these endpoints only show up on the temporary redirect-in-progress page, which users should never see a toast/error message on. Our current behaviour is that the frontend handles the error messages according to status code (we only get 500/401 errors) upon the user getting redirected to the login page.

File: src/routes/v2/sgidAuth.ts
Comment (alexanderleegs) [reply src/routes/v2/sgidAuth.ts:147]: The expiring/signing is to ensure that a user attempting to use the multiuser login actually has access to that user - at this point we actually are able to log in the user as any of their approved emails already. But since we don't know which user they want to use, this requires additional input from the user, which requires us to ensure that an attacker cannot simply call this endpoint pretending to be the user.

File: src/routes/v2/sgidAuth.ts
Comment (alexanderleegs) [reply src/routes/v2/sgidAuth.ts:180]: Not repeated! single use, separated to make it easier to read

File: src/routes/v2/sgidAuth.ts
Comment (seaerchin) [reply src/routes/v2/sgidAuth.ts:112]: hmm ok - i was hoping for the frontend to extract the error message out from teh backend isntead of it deriving the message from status code. 

is this a fast change? if not, fine iwth leaving as is and doing lkater

File: src/services/utilServices/SgidAuthService.ts
Comment (alexanderleegs) [reply src/services/utilServices/SgidAuthService.ts:104]: Split because i needed to cast earlier!

File: src/services/identity/CollaboratorsService.ts
Comment (seaerchin) [reply src/services/identity/CollaboratorsService.ts:161]: won't happen - email now is from sgid

File: src/routes/v2/sgidAuth.ts
Comment (seaerchin) [reply src/routes/v2/sgidAuth.ts:147]: don't they hit this point **only after** they pass authorization checks from sgid? 

the multi-user email selection flow seems to be covered below

Comment (seaerchin) [reply src/routes/v2/sgidAuth.ts:147]: oh ok just double checking - this authorizes and retrieves data from sgid then passes back to FE

user selects email on FE and hits the below endpoint, **which verifies the security token** to ensure identity

if this is the flow, then please resolve

**Review by seaerchin — APPROVED** [review]
lgtm if my assumption on the login flow for multi user is correct. otherwise, **please dismiss**

File: src/types/sgid.ts
Comment (seaerchin) [reply src/types/sgid.ts:11]: please do this before merge also

File: src/services/utilServices/SgidAuthService.ts
Comment (alexanderleegs) [reply src/services/utilServices/SgidAuthService.ts:97]: Hmm this should be handled by the `ResultAsync` wrapper right?

File: src/services/identity/CollaboratorsService.ts
Comment (alexanderleegs) [reply src/services/identity/CollaboratorsService.ts:161]: Yeah this is true - had a chat earlier and we decided that a quick fix for this would be to periodically go through the user table and remove users with no associated mobile numbers (since this means that have never made any cms edits before as you are blocked on doing anything until you fill in contact details)

File: src/errors/SgidErrors.ts
Comment (alexanderleegs) [reply src/errors/SgidErrors.ts:1]: https://github.com/isomerpages/isomercms-backend/pull/976/commits/ee9bbe1313676bbf9580e93594a03fd6d302bc00

File: src/routes/v2/sgidAuth.ts
Comment (alexanderleegs) [reply src/routes/v2/sgidAuth.ts:101]: officer details! https://github.com/isomerpages/isomercms-backend/pull/976/commits/ee9bbe1313676bbf9580e93594a03fd6d302bc00

File: src/routes/v2/sgidAuth.ts
Comment (alexanderleegs) [reply src/routes/v2/sgidAuth.ts:113]: Let's do this in a separate PR?

File: src/types/sgid.ts
Comment (alexanderleegs) [reply src/types/sgid.ts:12]: https://github.com/isomerpages/isomercms-backend/pull/976/commits/eacf951ad2a268bfa36dcf10b25b42a2ff1956db

File: src/routes/v2/sgidAuth.ts
Comment (alexanderleegs) [reply src/routes/v2/sgidAuth.ts]: 48a15e38

File: src/routes/v2/sgidAuth.ts
Comment (alexanderleegs) [reply src/routes/v2/sgidAuth.ts]: 48a15e38

File: src/routes/v2/sgidAuth.ts
Comment (alexanderleegs) [reply src/routes/v2/sgidAuth.ts]: b6434bda

**Pull request merged**
